nltools/stats.py

```diff
@@ -1191,3 +1192,175 @@ def procrustes(data1, data2):
+        yy = np.multiply(y_dist_cent,y_dist_cent).mean()
+    
+    # 3 compute covariances and variances  
+    cov = np.sqrt(xy)
+    var_x = np.sqrt(xx)
+    var_y = np.sqrt(yy)
+
+    # 4 Normalize to get correlation
+    denom = np.sqrt(xx * yy)
+    if denom > 0:
+        r2 = xy / denom
+    else:
+        r2 = 0
+    cor = np.sqrt(np.abs(r2)) # because u-centering can produce negative dcors
+    
+    out = {}
+    out['d_correlation'] = cor
+
+    if return_covs:
+        out['d_covariance'] = cov
+        out['x_var'] = var_x
+        out['y_var'] = var_y
+
+    if bias_corrected:
+        dof = (adjusted_n / 2) - 1
+        t = np.sqrt(dof) * (r2 / np.sqrt(1 - r2**2))
+        p = 2*(1-t_dist.cdf(np.abs(t), dof))
+        out['t'] = t
+        out['p'] = p
+        out['df'] = dof
+
+    return out
+
+def procrustes_similarity(mat1,mat2,n_permute=5000,tail=2,n_jobs=-1,random_state=None):
+    """ Use procrustes super-position to perform a similarity test between 2 matrices. Matrices need to match in size on their first dimension only, as the smaller matrix on the second dimension will be padded with zeros. After aligning two matrices using the procrustes transformation, use the computed disparity between them (sum of squared error of elements) as a similarity metric. Shuffle the rows of one of the matrices and recompute the disparity to perform inference (Peres-Neto & Jackson, 2001). Note: by default this function reverses disparity to treat it like a *similarity* measure like correlation, rather than a distance measure like correlation distance, i.e. smaller values mean less similar, larger values mean more similar.
+    
+    Args:
+        mat1 (ndarray): 2d numpy array; must have same number of rows as mat2
+        mat2 (ndarray): 1d or 2d numpy array; must have same number of rows as mat1
+        n_permute (int): number of permutation iterations to perform
+        tail (int): either 1 for one-tailed or 2 for two-tailed test; default 2
+        n_jobs (int): The number of CPUs to use to do permutation; default -1 (all)
+    
+    Returns:
+        similarity (float): similarity between matrices bounded between 0 and 1
+        pval (float): permuted p-value
+
+    """
+   
+    if mat1.shape[0] != mat2.shape[0]:
+        raise ValueError('Both arrays must match on their first dimension')
+    
+    random_state = check_random_state(random_state)
+    
+    # Make sure both matrices are 2d and the same dimension via padding
+    if len(mat1.shape) < 2:
+        mat1 = mat1[:,np.newaxis]
+    if len(mat2.shape) < 2:
+        mat2 = mat2[:,np.newaxis]
+    if mat1.shape[1] > mat2.shape[1]:
+        mat2 = np.pad(mat2,((0,0),(0, mat1.shape[1] - mat2.shape[1])),'constant')
+    elif mat2.shape[1] > mat1.shape[1]:
+        mat1 = np.pad(mat1,((0,0),(0, mat2.shape[1] - mat1.shape[1])),'constant')
+    
+     _,_,sse = procrustes(mat1,mat2)
+    sse = 1 - sse # flip to similarity measure
```

Comment:
are you sure that sum of squared error is scaled between 0 and 1?  This seems strange.  I think it's okay to leave it in distance.  There is a function in adjacency to approximate similarity from distance and the reverse (works for euclidean distance too)

Reply:
I thought so based on the literature but it looks like scipy’s implementation might not be. Basically need to divide sse outputted here by sse of centered X matrix so it becomes a ratio. Thats the matlab approach and approach in the literature. Ill update and check it.